torch/_decomp/decompositions.py

```diff
@@ -2065,6 +2065,154 @@ def upsample_nearest3d(
     return result
 
 
+def gather_params(params, has_biases, has_projections):
+    if has_biases and has_projections:
+        group_size = 5
+    elif has_biases:
+        group_size = 4
+    elif has_projections:
+        group_size = 3
+    else:
+        group_size = 2
+
+    assert len(params) % group_size == 0, len(params)
+    return [
+        tuple(params[i : i + group_size]) for i in range(0, len(params), group_size)
+    ]
+
+
+def params_hiddens(params, hiddens, i, bidirectional):
+    if bidirectional:
+        cur_params, cur_hidden = params[2 * i], hiddens[2 * i]
+        bidir_params, bidir_hidden = params[2 * i + 1], hiddens[2 * i + 1]
+    else:
+        cur_params, cur_hidden = params[i], hiddens[i]
+        bidir_params, bidir_hidden = None, None
+
+    return cur_params, cur_hidden, bidir_params, bidir_hidden
+
+
+def one_layer_rnn(inp, hidden, params, has_biases, nonlinearity, reverse=False):
+    ih_weight = params[0]
+    hh_weight = params[1]
+    ih_bias = params[2] if has_biases else None
+    hh_bias = params[3] if has_biases else None
+
+    precomputed_input = F.linear(inp, ih_weight, ih_bias)
+    precomputed_input = precomputed_input.flip(0) if reverse else precomputed_input
+    cur_hidden = hidden
+    step_output = []
+    for inp in precomputed_input:
+        cur_hidden = nonlinearity(F.linear(cur_hidden, hh_weight, hh_bias) + inp)
+        step_output.append(cur_hidden)
+
+    out = torch.stack(step_output, 0)
+
+    return out, cur_hidden
+
+
+def _rnn_helper(
+    input,
+    hidden,
+    params,
+    has_biases,
+    num_layers,
+    dropout,
+    train,
+    bidirectional,
+    batch_first,
+    layer_fn,
+):
+    input = input.transpose(0, 1) if batch_first else input
+    final_hiddens = []
+
+    for i in range(num_layers):
+        cur_params, cur_hidden, bidir_params, bidir_hidden = params_hiddens(
+            params, hidden, i, bidirectional
+        )
+        dropout = dropout if (train and num_layers < i - 1) else 0.0
+        fwd_inp, fwd_hidden = layer_fn(input, cur_hidden, cur_params, has_biases)
+        final_hiddens.append(fwd_hidden)
+
+        if bidirectional:
+            bwd_inp, bwd_hidden = layer_fn(
+                input, bidir_hidden, bidir_params, has_biases, reverse=True
+            )
+            bwd_inp = bwd_inp.flip(0)
+            final_hiddens.append(bwd_hidden)
+
+        if bidirectional:
+            input = torch.cat([fwd_inp, bwd_inp], fwd_inp.dim() - 1)
+        else:
+            input = fwd_inp
+
+        if dropout != 0 and train and i < num_layers - 1:
+            input = torch.dropout(input, dropout, train=True)
+
+    input = input.transpose(0, 1) if batch_first else input
+    return input, final_hiddens
+
+
+@aten.rnn_tanh.input.py_impl(DispatchKey.CompositeImplicitAutograd)
+@aten.rnn_tanh.input.py_impl(DispatchKey.Autograd)
+def rnn_tanh_input(
```

Comment:
What are the downstream ramifications of this change? e.g. does this cause PT2 to use this decomposition?

Reply:
Yes this causes PT2 to use the decomposition. From some other issue I thought that's what we want with the theory that dynamo/inductor would re-coop performance (also since I doubt we would be able to do symbolic shapes through RNNs since we don't know the buffer size)

But since those are closed, I would be happy to make these post autograd decompositions only instead of pre autograd. I don't really like the idea of having another set of decompositions for expanded weights only in case someone needs to add this to PT2 later (it's kinda hard to discover) but I'm open hear other opinions

Reply:
Is there a way to separate out the "adding to PT2" from this PR?

> From some other issue I thought that's what we want with the theory that dynamo/inductor would re-coop performance (also since I doubt we would be able to do symbolic shapes through RNNs since we don't know the buffer size)

Do you have a link to the issues and can we ask the folks cc'ed there for an opinion? In theory dynamo/inductor can re-coop the performance if we decompose RNNs. In practice, there's a lot they need to do to the replicate the RNN optimizations on CUDA; the cudnn kernels are highly optimized. I'd want to get some more eyes on this from people who work on PT2 perf (@ngimel?)

Reply:
> Do you have a link to the issues and can we ask the folks cc'ed there for an opinion?

https://github.com/pytorch/functorch/issues/586
Animesh is out and I'm not certain if this is the right function but it looks like @ezyang might have just disallowed dynamo from capturing RNNs at all

Reply:
If dynamo doesn't capture RNNs then my concerns are alleviated

Reply:
Well, I mean, I just caused us to hit `set_`, which means they get graph broken. RNNs are complicated because of the weight sharing thing, I am pretty doubtful inductor knows how to deal with it.